locked-asset/proxy_dex/src/proxy_common.rs

```diff
@@ -8,20 +8,81 @@ use crate::{
     wrapped_lp_attributes::WrappedLpTokenAttributes,
 };
 
-/// common interface for both old and new locked token factory
-pub mod locked_token_factory {
-    elrond_wasm::imports!();
-
-    #[elrond_wasm::proxy]
-    pub trait LockedTokenFactory {
-        #[payable("*")]
-        #[endpoint(mergeTokens)]
-        fn merge_tokens(&self) -> EsdtTokenPayment;
-    }
+pub static FACTORY_MERGE_TOKENS_ENDPOINT_NAME: &[u8] = b"mergeTokens";
+pub static INVALID_PAYMENTS_ERR_MSG: &[u8] = b"Invalid payments";
+
+pub struct LockedUnlockedTokenRefPair<'a, M: ManagedTypeApi> {
+    pub locked_token_ref: &'a EsdtTokenPayment<M>,
+    pub unlocked_token_ref: &'a EsdtTokenPayment<M>,
+}
+
+pub struct BaseAssetOtherTokenRefPair<'a, M: ManagedTypeApi> {
+    pub base_asset_token_ref: &'a EsdtTokenPayment<M>,
+    pub other_token_ref: &'a EsdtTokenPayment<M>,
 }
 
 #[elrond_wasm::module]
-pub trait ProxyCommonModule: token_send::TokenSendModule {
+pub trait ProxyCommonModule {
+    fn require_exactly_one_locked<'a>(
+        &self,
+        first_payment: &'a EsdtTokenPayment,
+        second_payment: &'a EsdtTokenPayment,
+    ) -> LockedUnlockedTokenRefPair<'a, Self::Api> {
+        let token_mapper = self.locked_token_ids();
+        let first_is_locked = token_mapper.contains(&first_payment.token_identifier);
+        let second_is_locked = token_mapper.contains(&first_payment.token_identifier);
+
+        if first_is_locked {
+            require!(!second_is_locked, INVALID_PAYMENTS_ERR_MSG);
+
+            LockedUnlockedTokenRefPair {
+                locked_token_ref: first_payment,
+                unlocked_token_ref: second_payment,
+            }
+        } else {
+            require!(second_is_locked, INVALID_PAYMENTS_ERR_MSG);
+
+            LockedUnlockedTokenRefPair {
+                locked_token_ref: second_payment,
+                unlocked_token_ref: first_payment,
+            }
+        }
+    }
+
+    fn require_exactly_one_base_asset<'a>(
+        &self,
+        first_payment: &'a EsdtTokenPayment,
+        second_payment: &'a EsdtTokenPayment,
+    ) -> BaseAssetOtherTokenRefPair<'a, Self::Api> {
+        let base_asset_token_id = self.asset_token_id().get();
+        let is_first_token = first_payment.token_identifier == base_asset_token_id;
+        let is_second_token = second_payment.token_identifier == base_asset_token_id;
+
+        if is_first_token {
+            require!(!is_second_token, INVALID_PAYMENTS_ERR_MSG);
```

Comment:
Is there a more specific error message than `INVALID_PAYMENTS_ERR_MSG`?

Reply:
The message could be something like "Expected at least one token as base asset", but keeping this somewhat generic message helps with code size, since instead of having many different static strings, we have a single one, used from multiple places.